please review my code maciek

maro243 opened the pull request.

maro243 committed:
all tests pass

maciekgajewski commented on course3/homework-string/my_string.h:
```diff
@@ -1,6 +1,63 @@
+MyString::MyString(const char* inchar){
+	int n = std::strlen(inchar);
+	data = std::unique_ptr<char[]>(new char[n]);
+ 	std::strcpy(data.get(), inchar);
```
You forgot about the zero-terminator in the string. `"abc"` requires 4-character buffer: a, b, c, \0, but `std::strlen("abc")` will return 3. If you decide to store null-terminated string in your buffer, you need to remember to add one character to store the terminator.

`stdcpy` copies the string and null-terminator. Because your buffer is too small, it writes over the memory you don't own. It should show-up in valgrind.

maciekgajewski replied:
Because you fail to add 1 to the buffer size, interesting things are happening with empty strings - you allocate 0 bytes.

maciekgajewski commented on course3/homework-string/my_string.h:
```diff
@@ -1,6 +1,63 @@
+	char& operator[](const int) const; // const indexing operator
+};
+
+MyString::MyString():MyString(""){};
```
This may be correct, but you perform unnecessary allocation only store empty string.

maciekgajewski commented on course3/homework-string/my_string.h:
```diff
@@ -1,6 +1,63 @@
+char& MyString::operator[](const int i) const{
+	char a = data.get()[i];
+	char &ref_a = a;
+	return ref_a;
```
whoa! You are returning a pointer (reference) to local variable. The variable is no longer valid when the function returns. Any access to the reference is a violation, and should be reported by valgrind.

I know why you did it; it has to do with const-correctness. Please, compare the signature of your operator[] with the ones of `std::string`

maciekgajewski commented:
Please, run the program trough valgrind, or another similar tool. There is few invalid operations there.

maro243 committed:
responding to feedback

maro243 committed:
using valgrind I found some memory leaks, cleaned them up!

maro243 committed:
merging in new test suite

maro243 committed:
implemented c_str method

maro243 committed:
tonight's progress, copy-assignment operator and move semantics`

maro243 committed:
added non rvalue copy-assignment operator. However there is a memory leak

maro243 committed:
all tests pass up until and including operator<

maro243 commented:
Hi Maciek, I've implemented most of the methods without any memory leaks.
Sorry for being a bit behind with homework but I will get through everything at my own pace, I am enjoying the course!

maciekgajewski commented on course3/homework-string/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -std=c++11")
```
We use C++14

maro243 replied:
yeah I'm using gcc 4.8.4 (comes with ubuntu 14), I only discovered last night that the right flag for c++14 is -std=c++1y

maciekgajewski commented on course3/homework-string/my_string.cc:
```diff
@@ -0,0 +1,113 @@
+# include <iostream>
+
+MyString::MyString(){
+//	std::cout << "default constructor" << std::endl;
```
Remove the comment-out code once you don't use it anymore

maciekgajewski commented on course3/homework-string/my_string.cc:
```diff
@@ -0,0 +1,113 @@
+};
+
+int MyString::size() const{
+	if (data == 0)
```
Please, don't compare pointer to int literal, this makes code confusing ("is 'data' an integer?). Either compare it to `nullptr`, or just take advantage of implicit boolean conversion.

maciekgajewski commented on course3/homework-string/my_string.cc:
```diff
@@ -0,0 +1,113 @@
+
+bool operator==(const MyString &s1, const MyString &s2){
+//	std::cout << "comparison operator " << s1.data.get()<< "==" << s2.data.get() << "?" << std::endl;
+	return std::strcmp(s1.data.get(), s2.data.get()) == 0;
```
What if data is null on any of the strings?

maciekgajewski commented on course3/homework-string/my_string.cc:
```diff
@@ -0,0 +1,113 @@
+}
+
+const char* MyString::c_str() const{
+	return data.get();
```
What if data is null?

maro243 replied:
what should the desired behaviour be, return std::nullptr or const char\* "" ?

maciekgajewski replied:
Follow the std::strings's behaviour:
http://en.cppreference.com/w/cpp/string/basic_string/c_str

maciekgajewski commented on course3/homework-string/my_string.cc:
```diff
@@ -0,0 +1,113 @@
+bool MyString::operator<(const MyString& s2){
+	if (this->empty()and s2.empty()){return false;}
+	if (this->empty()and !s2.empty()){return true;}
+	if (!this->empty()and s2.empty()){return false;}
```
If your `c_str()` worked correctly you could reduce it to single strcmp call

maro243 committed:
responding to feedback
deleting comments
using nullptr instead of 0
c_str returns "" properly, enabled be to reduce spaghetti in operator+ and operator<

maro243 committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	course3/homework-string/test.cc

maro243 committed:
starting unique ptr assignment

maro243 commented on course3/homework-uniqueptr/test.cc:
```diff
@@ -0,0 +1,20 @@
+	test<std::unique_ptr<char[]>>();
+	std::cout<<std::endl;
+	std::cout<<"test MyUniquePtr"<<std::endl;
+	test<MyUniquePtr<char[]>>();
```
Hi Maciek, looking for some advice

at compile time I'm getting the following which works for std::unique_ptr
test.cc:9:23: error: no matching function for call to ‘MyUniquePtr<char []>::MyUniquePtr(char_)
The type I instantiate MyUniquePtr with is char [],  the call to (new char[5]) returns type char \* so I see why this is happening, but I can't think of any solution, I would hope that the compiler could kind of ignore the difference between char_ and char[]. My suspicion is maybe there is no neat solution for arrays and I would need to add specialisation for that case, I think I remember you saying something about that.

On another note, I was trying to structure my test function to just take the class template MyUniquePtr, and delved into the concept of class template templates. Couldn't really get it working so I just decided to pass the instantiated UniquePtr<char[]> to the test function template instead. Would you be able to off any advice on how to set up tests for template classes?

maciekgajewski replied:
1. Don't count on compiler ignoring the difference between `char` and `char[]`. They are _always_ different types, and, sadly, in different contexts the [] means different thing.
   std::unique_ptr provides _specialization_ for [] types, which stores arrays instead of single objects. If you don't know how to create such a specialization, please, look for inspiration in the code of std::unique_ptr.

Please note that unique_ptr<cha> and unique_ptr<char[]> are completely different types, with different interface, that do different things.
1. Template templates are nightmare fuel. The solution you used is a valid one.
